src/adapter/src/catalog.rs

```diff
@@ -2743,14 +2698,31 @@ impl Default for DefaultPrivileges {
 impl DefaultPrivileges {
     /// Add a new default privilege into the set of all default privileges.
     fn grant(&mut self, object: DefaultPrivilegeObject, privilege: DefaultPrivilegeAclItem) {
+        if privilege.acl_mode.is_empty() {
+            return;
+        }
+
         let privileges = self.privileges.entry(object).or_default();
-        if let Some(default_privilege) = privileges
-            .into_iter()
-            .find(|default_privilege| default_privilege.grantee == privilege.grantee)
-        {
+        if let Some(default_privilege) = privileges.get_mut(&privilege.grantee) {
             default_privilege.acl_mode |= privilege.acl_mode;
         } else {
-            privileges.push(privilege);
+            privileges.insert(privilege.grantee, privilege);
+        }
+    }
+
+    /// Revoke a default privilege from the set of all default privileges.
+    fn revoke(&mut self, object: &DefaultPrivilegeObject, privilege: &DefaultPrivilegeAclItem) {
+        if let Some(privileges) = self.privileges.get_mut(object) {
+            if let Some(default_privilege) = privileges.get_mut(&privilege.grantee) {
+                default_privilege.acl_mode =
+                    default_privilege.acl_mode.difference(privilege.acl_mode);
+                if default_privilege.acl_mode.is_empty() {
+                    privileges.remove(&privilege.grantee);
+                }
+            }
+            if privileges.is_empty() {
```

Comment:
An empty set means revoke all?

Reply:
No, this is to cover the scenario where we've revoked all `DefaultPrivilegeAclItem`s that match a given `DefaultPrivilegeObject`, then we clean up it's entry in the map.